refactor: EnumでOpenStructの利用をやめる

jk-es335 opened the pull request:
## 概要
- ruby3のOpenStruct非推奨に伴いEnumでの利用を廃止
    - インクルードしたモジュールに、`Enum::Entry`のサブクラスを動的に追加する
    - Enum::Entryの同値判定は、key / order / nameが一致していることとする（OpenStruct利用時と同様の挙動）
    - method_missingを撤廃（パフォーマンス観点）
    - enums_namesを撤廃
    - proc/lambdaの対応（クラスロード時の定義では要件を満たさない場合に対応）
- enum.rbのspecを追加
- simplecovを追加
- versionを0.3.0へ
- eval_erbではまだOpenStructを利用している
- コード自動生成スクリプトではエラーが発生しないことを確認済み
- values.pluck(:hoge)のように`pluck`メソッドが使えることを確認済み（`[]`メソッドを追加した）

## benchmark


| Version | Access Type | Iterations | User Time | System Time | Total Time | Real Time |
|---------|-------------|------------|-----------|-------------|------------|-----------|
| OpenStruct Version | enum access | 1000 | 0.000469 | 0.000029 | 0.000498 | 0.000497 |
| This Version | enum access | 1000 | 0.000126 | 0.000023 | 0.000149 | 0.000191 |
| OpenStruct Version | enum access | 10000 | 0.004125 | 0.000075 | 0.004200 | 0.004227 |
| This Version | enum access | 10000 | 0.001043 | 0.000003 | 0.001046 | 0.001054 |
| OpenStruct Version | enum access | 100000 | 0.032144 | 0.000998 | 0.033142 | 0.033224 |
| This Version | enum access | 100000 | 0.009313 | 0.000008 | 0.009321 | 0.009334 |
| OpenStruct Version | method access | 1000 | 0.000319 | 0.000001 | 0.000320 | 0.000319 |
| This Version | method access | 1000 | 0.000239 | 0.000002 | 0.000241 | 0.000246 |
| OpenStruct Version | method access | 10000 | 0.003319 | 0.000102 | 0.003421 | 0.003429 |
| This Version | method access | 10000 | 0.002284 | 0.000003 | 0.002287 | 0.002292 |
| OpenStruct Version | method access | 100000 | 0.028652 | 0.000209 | 0.028861 | 0.028960 |
| This Version | method access | 100000 | 0.019851 | 0.000020 | 0.019871 | 0.019947 |

jk-es335 committed:
refactor: EnumでOpenStructの利用をやめる

jk-es335 commented on lib/machiiro/ruby/support/helper/enum.rb:
```diff
@@ -78,21 +81,60 @@ def index_of(key)
+      end
+
+      def as_json(*)
+        @hash.as_json(only: %i[key name order lower_name])
```
ここを固定キーにしていいのか気になっています

TanaShun0804 replied:
んー自分もちょっと判断つかずですね...

jk-es335 replied:
現行の確認をしてfield全てだしていましたのでonlyはなくしました

TanaShun0804 reviewed (approved):
現行挙動維持できてますし問題ないかと思います！LG!

jk-es335 committed:
test: spec修正

jk-es335 committed:
refactor: as_jsonでフィールド絞り込みをやめる（現行の挙動を確認した）

TanaShun0804 commented on spec/machiiro/ruby/support/helper/enum_spec.rb:
```diff
@@ -0,0 +1,290 @@
+        end
+      end
+
+      let(:enum2) do
+        Module.new do
+          include MachiiroSupport::Enum
+
+          enums_ordinal [:AUTO, label: ''],
+                        [:ENTRY, label: '入館'],
+                        [:EXIT, label: '退館']
+        end
+      end
```
いまさらですがここ不要ですかね？

jk-es335 replied:
ありがとうございます！消します！

jk-es335 replied:
削除しました！

jk-es335 committed:
test: 不要なletを削除

iwazzer commented on lib/machiiro/ruby/support/helper/enum.rb:
```diff
@@ -37,21 +37,24 @@ def enums_string(*_enums)
           super
         end
       end
-      ruby2_keywords(:method_missing) if respond_to?(:ruby2_keywords, true)
```
確かに、Enum は、この対応不要でしたね。ありがとうございます。

iwazzer commented on lib/machiiro/ruby/support/helper/enum.rb:
```diff
@@ -78,21 +81,60 @@ def index_of(key)
-          end
-
-          q_method = "#{name.downcase}?"
-          q_method = "_#{q_method}" if OpenStruct.method_defined?(q_method)
```
hacomono の Master::RecordFormatFillType.BLANK が 2.7 と 3.0 で逆になり、結果 present? を呼ぶところで逆に判別されて spec が落ちてたので、明示的に BLANK の値を blank? で参照していたところがなさそうだったため、 _blank? に置き換えて、blank? は Object#blank? が返るようにしました。
メモによると失敗してたのはココ
https://github.com/machiiro/hacomono/blob/v3.0%2Fproduction/api/layer/helper/migration/import_text_row.rb#L119

今回の修正では OpenStruct じゃないから大丈夫とかなのかしら？🤔

jk-es335 replied:
おーーみてみます！

jk-es335 replied:
テストケースを追加しました！
結論として2.7の時と同一の挙動になります

iwazzer reviewed (approved):
気になった点をコメントしました(｡･ω･｡)ﾉ

jk-es335 committed:
test: blank?やpresent?などキーと同一名称のケースを追加

jk-es335 committed:
refactor: remove method_missing

jk-es335 commented:
@TanaShun0804 @iwazzer 
すみません、method_missingやproc/lambda対応をしましたので再度レビューをお願いいたします

jk-es335 committed:
refactor: use underscore method for instance variable name

jk-es335 commented on lib/machiiro/ruby/support/helper/enum.rb:
```diff
@@ -1,57 +1,44 @@
-          end
-          hash[:lower_name] = hash[:name].downcase
-          hash = add_judgement_items(_enums, e, hash)
+      def generate_ivar_name(name)
```
to_hメソッドでキャッシュを利用するためにインスタンス変数が必要になります。
rubyの仕様ではインスタンス変数名には"?"や"-"が使えないので変換をしています。

jk-es335 commented on lib/machiiro/ruby/support/helper/enum.rb:
```diff
@@ -62,8 +49,11 @@ def string?
 
       def values
-        enums
+        @values ||= names.each_with_object([]) do |name, entries|
```
Array<T\<extend Entry\>>になります

jk-es335 commented on lib/machiiro/ruby/support/helper/enum.rb:
```diff
@@ -76,23 +66,135 @@ def index_of(key)
         values.index { |v| v.key == key }
       end
 
+      def has?(name)
```
`enums_names.has_key?`
の代替用です

jk-es335 commented on lib/machiiro/ruby/support/helper/enum.rb:
```diff
@@ -76,23 +66,135 @@ def index_of(key)
           end
+        end
+
+        const_set(clazz_name, clazz)
```
ここでconst_setをすることで、無名クラスがクラス名を持つことになります

jk-es335 commented on lib/machiiro/ruby/support/helper/enum.rb:
```diff
@@ -76,23 +66,135 @@ def index_of(key)
-          hash[q_method] = _e == e
+        # prohibit to create instance of Entry
+        clazz.class_eval do
+          private_class_method :new
```
Entryのインスタンス化は禁止します

jk-es335 commented on lib/machiiro/ruby/support/helper/enum.rb:
```diff
@@ -76,23 +66,135 @@ def index_of(key)
+        @lower_name = name.downcase
+      end
+
+      def [](key)
```
enums.pluck(:xxx)
とやると内部で[]の呼び出しが走るので互換性のために定義しています

jk-es335 commented on lib/machiiro/ruby/support/helper/enum.rb:
```diff
@@ -76,23 +66,135 @@ def index_of(key)
+
+            define_method(k) do
+              # set value to instance variable for Entry#to_h
+              instance_variable_set(ivar, v.respond_to?(:call) ? v.call : v)
```
このreaderメソッドが呼ばれたタイミングでproc/lambdaの評価を行います

shouhei-matsuo-0228 replied:
これがproc/lambdaは遅延評価に使えると言っていたやつですね〜。すごいー。

jk-es335 committed:
refactor: use each_with_index instead od map

jk-es335 committed:
refactor: use to_sym for `names`

iwazzer commented on lib/machiiro/ruby/support/helper/enum.rb:
```diff
@@ -76,23 +66,135 @@ def index_of(key)
+        end
+      end
+      alias eql? ==
+      alias equal? ==
```
これは置き換える必要あるんですかね？🤔

iwazzer replied:
hash 置き換えてるから同義か

iwazzer commented on lib/machiiro/ruby/support/version.rb:
```diff
@@ -1,3 +1,3 @@
 module MachiiroSupport
-  VERSION = "0.2.0"
+  VERSION = "0.3.0"
```
この変更、かなり有用だし全部の環境が Ruby 3 になるの待ってるのがもったいないっすね😅
0.1.1 とかでバックポートして先に入れて、0.3（0.2.1 でも）用は別途とかにする？ｗ

iwazzer replied:
いやー、でも両方の確認とるのめんどくさいか😅

iwazzer replied:
0.3.0 のタグを打って hacomono 側も更新すればいいだけか💡

iwazzer reviewed (approved):
良さそうです(•̀ᴗ•́)و ̑̑

TanaShun0804 reviewed (approved):
レビューできる技量が足りませんでしたがspec見る限りは問題ないと思います！

jk-es335 commented:
> レビューできる技量が足りませんでしたがspec見る限りは問題ないと思います！

いえいえ、お忙しい中ありがとうございました！

jk-es335 committed:
refactor: use hash for value_of

jk-es335 committed:
refactor: change scope of Entry for expansion

shouhei-matsuo-0228 reviewed (approved):
LGTM！

jk-es335 committed:
chore: rollback scope of Entry

iwazzer committed:
build: HCREEN-XXX Bump ruby rails

- ERB.new の deprecated オププションを除去

iwazzer committed:
build: HCREEN-XXX Bump ruby rails

- YAML.load_file のデフォルトが safe_load_file に変更されたので、明示的に unsafe_load_file を使えるオプションを追加

iwazzer committed:
build: HCREEN-XXX Bump ruby rails

- rails のバージョン依存を外す

iwazzer committed:
build: HCREEN-141 Bump ruby rails

- use public_send

iwazzer committed:
build: HCREEN-141 Bump ruby rails

- Bump version.rb

iwazzer committed:
Merge pull request #13 from machiiro/build/HCREEN-141-bump-ruby-rails

build: for bump ruby rails
